lib/ddtrace/runtime/container.rb

```diff
@@ -0,0 +1,52 @@
+require 'ddtrace/ext/runtime'
+require 'ddtrace/runtime/cgroup'
+
+module Datadog
+  module Runtime
+    # For container environments
+    module Container
+      UUID_PATTERN = '[0-9a-f]{8}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{12}'.freeze
+      CONTAINER_PATTERN = '[0-9a-f]{64}'.freeze
+
+      POD_REGEX = /(pod)?(#{UUID_PATTERN})(?:.slice)?$/
+      CONTAINER_REGEX = /(#{UUID_PATTERN}|#{CONTAINER_PATTERN})(?:.scope)?$/
+
+      Descriptor = Struct.new(
+        :platform,
+        :container_id,
+        :task_uid
+      )
+
+      def self.platform
+        descriptor.platform
+      end
+
+      def self.container_id
+        descriptor.container_id
+      end
+
+      def self.task_uid
+        descriptor.task_uid
+      end
+
+      def self.descriptor
+        @descriptor ||= begin
+          Descriptor.new.tap do |descriptor|
+            Cgroup.descriptors.each do |cgroup_descriptor|
+              # Parse container data from cgroup descriptor
+              path = cgroup_descriptor.path
+
+              next if path.nil?
```

Comment:
We should probably also only consider a path if it containers a `CONTAINER_REGEX`.

Not sure I have an example I can share right away, but I remember seeing one where not all the path parts in the cgroup file were the same. This was for a non-containerized environment, but still makes me wonder if that is possible in a containerized environment too.

Should we be defensive here, or am I being overly cautious?

Reply:
I don't think your general instinct is wrong here; the JS algo I was translating from did actually iterate over every line, and if it had a match for any of these values, it would capture it. I thought I was being a little faster by just taking the first eligible, but maybe that's not sufficient and it should iterate over every line.

I guess what does it mean for the purposes of container tagging when you have control groups with different paths? I'm not sure what implications that has for this trace's container ID.

Reply:
Ah, here is the example I found where they were different:

```
11:cpuset:/
10:freezer:/
9:net_cls,net_prio:/
8:hugetlb:/
7:blkio:/user.slice/user-0.slice/session-14.scope
6:pids:/user.slice/user-0.slice/session-14.scope
5:memory:/user.slice/user-0.slice/session-14.scope
4:devices:/user.slice/user-0.slice/session-14.scope
3:cpu,cpuacct:/user.slice/user-0.slice/session-14.scope
2:perf_event:/
1:name=systemd:/user.slice/user-0.slice/session-14.scope
```

Reply:
again, not sure we care much here, because this is not a containerized environment so...

Reply:
I used my local `cgroup` as well, it looks similar. I can make sure there's an example of it in the container specs, to make sure it doesn't break.

Reply:
Okay, it was having troubles parsing this; worked on an earlier iteration but didn't update accordingly. Added examples and should be working now.